Output stacks (support for SVG visualization)

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48438 Output stacks (support for SVG visualization)**

Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

Commit:
Output stacks (support for SVG visualization)

Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">


[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">


[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">


[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

dzhulgakov commented on torch/autograd/profiler.py:
```diff
@@ -261,6 +261,21 @@ def export_chrome_trace(self, path):
             f.write("]")
 
+    def export_stacks(self, path: str, metric: str):
+        assert metric in ["self_cpu_time_total", "self_cuda_time_total"]
```
nicer error? like `ValueError` and with a good message?

dzhulgakov commented on torch/autograd/profiler.py:
```diff
@@ -504,6 +519,24 @@ def export_chrome_trace(self, path):
+          ./flamegraph.pl --title "CPU time" --countname "us." profiler.stacks > perf_viz.svg
+        """
+        self._check_finish()
+        assert metric in ["self_cpu_time_total", "self_cuda_time_total"]
+        assert self.function_events is not None
+        assert self.with_stack, "export_stacks() requires with_stack=True"
```
better error messages for all of these please

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]

Commit:
Update on "Output stacks (support for SVG visualization)"


Summary:
Outputting stacks in a format suitable for SVG vizualization
(e.g. with https://github.com/brendangregg/FlameGraph tool)

Test Plan:
python test/test_profiler.py -k test_export_stacks

e.g. resnet18 (note: actual SVG is interactive):

<img width="1193" alt="Screen Shot 2020-11-24 at 7 06 27 PM" src="https://user-images.githubusercontent.com/30845429/100178160-397f3500-2e88-11eb-81c4-34b19c5fcb87.png">

Differential Revision: [D25174270](https://our.internmc.facebook.com/intern/diff/D25174270)

[ghstack-poisoned]
